### DISCUSSION TITLE
ERROR: argument of AND must be type boolean, not type character varying

### DISCUSSION THREAD

**Discussion by devinat1** [sqlalchemy/sqlalchemy#8314] [Usage Questions] [answered]
I am posting this under discussions as I am new to SQLAlchemy.
I am getting the ERROR: argument of AND must be type boolean, not type character varying
This is my table:
```
Table("userinfo", self.metadata,
                        Column("sub", String, primary_key=True, nullable=False),
                        Column("email", String),
                        Column("alpaca_oauth_token", String),
                        Column("alpaca_id", String),
                        Column("gains", Integer),
                        Column("algo_selection", Boolean),
                        Column("first_name", String),
                        Column("last_name", String),
                        Column("stocks", JSON))
```

This is the code executed:
```
        with self.engine.begin() as conn:
            select_stmt = self.userinfo.select(self.userinfo.c.alpaca_oauth_token).where(self.userinfo.c.sub == sub)
            result = conn.execute(select_stmt)
            row = result.all()
```

Compiled sql statment:
```
SELECT userinfo.sub, userinfo.email, userinfo.alpaca_oauth_token, userinfo.alpaca_id, userinfo.gains, userinfo.algo_selection, userinfo.first_name, userinfo.last_name, userinfo.stocks
FROM userinfo
WHERE userinfo.alpaca_oauth_token AND userinfo.sub = %(sub_1)s]
```

If I changed the statement to: 
```
SELECT userinfo.sub, userinfo.email, userinfo.alpaca_oauth_token, userinfo.alpaca_id, userinfo.gains, userinfo.algo_selection, userinfo.first_name, userinfo.last_name, userinfo.stocks
FROM userinfo
WHERE userinfo.alpaca_oauth_token **IS NOT NULL** AND userinfo.sub = %(sub_1)s]
```

Then I get the desired output.
I am running the psycopg2-binary==2.9.3 driver along with SQLAlchemy==1.4.39 and engine future enabled on WSL2 Ubuntu 20.04 and Python 3.8.10.

If I create the same connection but instead create_engine with a sqlite in memory database or database file, then I get the desired output. The issue seems to be specifically with postgresql.

**Comment by devinat1** [comment]
Fixed by changing: `select_stmt = self.userinfo.select(self.userinfo.c.alpaca_oauth_token).where(self.userinfo.c.sub == sub)`
To: `select_stmt = select(self.userinfo.c.alpaca_oauth_token).where(self.userinfo.c.sub == sub)`

However, it would be nice to know why this change worked.

**Comment by zzzeek** [comment, marked as answer]
you are using an argument in [table.select()](https://docs.sqlalchemy.org/en/14/core/metadata.html?highlight=table+select#sqlalchemy.schema.Table.select) that's deprecated, which is a WHERE criteria.   the API that this looks like is:

```py
stmt = table.select(table.c.col == some_value)
```

in 2.0, `table.select()` accepts no arguments, that WHERE criteria being passed is a very old calling form.

what you mean to do here is:

```py
from sqlalchemy import select
select_stmt = select(self.userinfo.c.alpaca_oauth_token).where(self.userinfo.c.sub == sub)
```
